rotkehlchen/chain/evm/decoding/llamazip/decoder.py

```diff
@@ -0,0 +1,89 @@
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.chain.evm.structures import EvmTxReceiptLog
+    from rotkehlchen.history.events.structures.evm_event import EvmEvent
+    from rotkehlchen.user_messages import MessagesAggregator
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class LlamazipCommonDecoder(DecoderInterface, abc.ABC):
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+            router_addresses: tuple['ChecksumEvmAddress', ...],
+    ) -> None:
+        super().__init__(
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+        self.router_addresses = router_addresses
+
+    def _decode_swap(
+            self,
+            transaction: EvmTransaction,
+            decoded_events: list['EvmEvent'],
+            all_logs: list['EvmTxReceiptLog'],  # pylint: disable=unused-argument
+    ) -> list['EvmEvent']:
+        spend_event = receive_event = None
+        for event in decoded_events:
+            if (
+                (event.event_type == HistoryEventType.TRADE and event.event_subtype == HistoryEventSubType.SPEND) or  # noqa: E501
+                (event.event_type == HistoryEventType.SPEND and event.event_subtype == HistoryEventSubType.NONE)  # noqa: E501
+            ):
+                spend_event = event
+                spend_event.counterparty = CPT_LLAMAZIP
+                spend_event.event_type = HistoryEventType.TRADE
+                spend_event.event_subtype = HistoryEventSubType.SPEND
+                spend_event.notes = f'Swap {event.balance.amount} {event.asset.symbol_or_name()} in LlamaZip'  # noqa: E501
+            elif (
+                event.event_type == HistoryEventType.RECEIVE and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.address in self.router_addresses
+            ):
+                receive_event = event
+                receive_event.counterparty = CPT_LLAMAZIP
+                receive_event.event_type = HistoryEventType.TRADE
+                receive_event.event_subtype = HistoryEventSubType.RECEIVE
+                receive_event.notes = f'Receive {event.balance.amount} {event.asset.symbol_or_name()} as the result of a swap in LlamaZip'  # noqa: E501
+
+        if None not in (spend_event, receive_event):
+            maybe_reshuffle_events(
+                ordered_events=[spend_event, receive_event],
+                events_list=decoded_events,
+            )
+        else:
+            log.error(f'Failed to find LlamaZip swap events for {transaction}')
+
+        return decoded_events
+
+    # -- DecoderInterface methods
+
+    def addresses_to_counterparties(self) -> dict[ChecksumEvmAddress, str]:
```

Comment:
This is not your own fault as functionality may be missing but a note to you and @yabirgb and @OjusWiZard.

The `addresses_to_counterparties` is missing the ability to drill down into the information as much as we need in this (and I guess) other particular cases we use it.

Ad discussed in discord here we would want not only `to` address being the router_address but also the input data starting with the proper function selector/s.

But here even if you just send a no value transfer to the router address (I know silly example -- but you get my point), it will get in this logic.

We need a special form of `addresses_to_counterparties` where the input data are also checked.